GitUI/UserControls/BlameControl.cs

```diff
@@ -641,8 +646,57 @@ private void blamePreviousRevisionToolStripMenuItem_Click(object sender, EventAr
+            int calculatedLine = selectedLine;
+
+            if (selectedLine != 1 && parentId != null)
+            {
+                try
+                {
+                    string currentLineText = _lastBlameLine.Text.Split("\n").First();
+
+                    // Get the git diff of changes introduced by the blamed selected commit
+                    // *without context lines* to make starting line value more accurate and possibly have more distincts chunks!
+                    GitArgumentBuilder args = new("diff") { $"-U0", parentId, blameInfo.selectedRevision.ObjectId, "--", blameInfo.filename };
+
+                    string diffOutput = UICommands.Module.GitExecutable.GetOutput(args, cache: GitModule.GitCommandCache);
+
+                    // Find the good chunk of code with current line number (starting from the end)
+                    foreach (string chunk in diffOutput.Split("\n@@").Skip(1).Reverse())
+                    {
+                        Match match = ChunkHeaderRegex().Match("@@" + chunk);
+                        if (!match.Success)
+                        {
+                            Trace.WriteLine("Blame previous commit: chunk/regex not matching\n@@" + chunk);
+                            continue;
+                        }
+
+                        int currentChunkStartingLine = int.Parse(match.Groups["CurrentLineNumber"].Value);
+                        if (currentChunkStartingLine < selectedLine)
+                        {
+                            // Calculate the offset thanks to the diff chunk header
+                            int previousChunkStartingLine = int.Parse(match.Groups["PreviousLineNumber"].Value);
+                            int lineRemovedCount = match.Groups["LineRemovedCount"].Success ? int.Parse(match.Groups["LineRemovedCount"].Value) : 1;
+                            int lineAddedCount = match.Groups["LineAddedCount"].Success ? int.Parse(match.Groups["LineAddedCount"].Value) : 1;
+
+                            // calculate the new position based on changed made above in the file
+                            // and changes made in this chunk
+                            calculatedLine = Math.Max(previousChunkStartingLine,
+                                selectedLine - currentChunkStartingLine + previousChunkStartingLine - lineAddedCount + lineRemovedCount);
+                            break;
+                        }
+                    }
+                }
+                catch (Exception ex)
+                {
+                    Trace.WriteLine("Blame previous commit: error when evaluating a better line number. Exception:" + ex);
+                }
+            }
+
+            _clickedBlameLine = new GitBlameLine(_lastBlameLine.Commit, _lastBlameLine.OriginLineNumber, calculatedLine, "Dummy Git blame line used only to store the good 'originLineNumber' value to display and select it");
+            BlameRevision(parentId, blameInfo.filename);
```

Comment:
Please extract it into own method. Maybe even create `GitBlameParser : IGitBlameParser`, so that it can be easily tested/injected/swapped out.

And add tests please.

Reply:
Done. Have a look if it suits you.

I have added an integration test on the GitExtensions repository on the case described in the commit message.